### PR TITLE
Eth JSON-RPC: populate reward in eth_feeHistory

### PR DESCRIPTION
**Pull request by ychiaoli18**
## Related Issues
https://github.com/filecoin-project/lotus/issues/10236

## Checklist

Before you mark the PR ready for review, please make sure that:

- [ ] Commits have a clear commit message.
- [ ] PR title is in the form of of `<PR type>: <area>: <change being made>`
  - example: ` fix: mempool: Introduce a cache for valid signatures`
  - `PR type`: fix, feat, build, chore, ci, docs, perf, refactor, revert, style, test
  - `area`, e.g. api, chain, state, market, mempool, multisig, networking, paych, proving, sealing, wallet, deps
- [ ] New features have usage guidelines and / or documentation updates in
  - [ ] [Lotus Documentation](https://lotus.filecoin.io)
  - [ ] [Discussion Tutorials](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] Tests exist for new functionality or change in behavior
- [ ] CI is green

### CODE REVIEW

**Review by magik6k — APPROVED** [review]

File: chain/types/ethtypes/eth_transactions.go [hunk 1]
```diff
@@ -43,6 +43,20 @@ type EthTx struct {
 	S                    EthBigInt   `json:"s"`
 }
 
+func (tx *EthTx) Reward(blkBaseFee *big.Int) (*EthBigInt) {
```
Comment (magik6k) [inline chain/types/ethtypes/eth_transactions.go:46]: linter complains

```suggestion
func (tx *EthTx) Reward(blkBaseFee *big.Int) *EthBigInt {
```

**Review by magik6k — COMMENTED** [review]
Itests fail
```
    eth_fee_history_test.go:55: 
        	Error Trace:	/home/circleci/project/itests/eth_fee_history_test.go:55
        	Error:      	Not equal: 
        	            	expected: 17
        	            	actual  : 16
        	Test:       	TestEthFeeHistory
    blockminer.go:362: shutting down mining
    blockminer.go:362: shutting down mining
--- FAIL: TestEthFeeHistory (9.00s)
```

File: node/impl/full/eth.go [hunk 2]
```diff
@@ -2128,3 +2160,50 @@ func parseEthTopics(topics ethtypes.EthTopicSpec) (map[string][][]byte, error) {
...
+	sort.Sort(txGasRewards)
+
+	var idx int
+	var sum uint64
+	for i, percentile := range rewardPercentiles {
+		threshold := uint64(float64(totalGasUsed) * percentile / 100)
+		for sum < threshold && idx < len(txGasRewards)-1 {
+			sum += txGasRewards[idx].gas
+			idx++
+		}
```
Comment (raulk) [inline node/impl/full/eth.go:2188]: If I'm reading this correctly, this fails to consider the last txGasRewards.

File: chain/types/ethtypes/eth_transactions.go [hunk 3]
```diff
@@ -44,6 +44,14 @@ type EthTx struct {
 	S                    EthBigInt   `json:"s"`
 }
 
+func (tx *EthTx) Reward(blkBaseFee big.Int) EthBigInt {
+	availablePriorityFee := big.Sub(big.Int(tx.MaxFeePerGas), blkBaseFee)
+	if big.Cmp(big.Int(tx.MaxPriorityFeePerGas), availablePriorityFee) <= 0 {
+		return tx.MaxPriorityFeePerGas
+	}
+	return EthBigInt(availablePriorityFee)
+}
```
Comment (raulk) [inline chain/types/ethtypes/eth_transactions.go:53]: I rewrote this a little to make it nicer to read.

File: node/impl/full/eth.go [hunk 4]
```diff
@@ -2128,3 +2160,50 @@ func parseEthTopics(topics ethtypes.EthTopicSpec) (map[string][][]byte, error) {
...
+		rewards[i] = ethtypes.EthBigIntZero
+	}
+
+	if len(txGasRewards) == 0 {
+		return rewards, totalGasUsed
+	}
+
+	sort.Sort(txGasRewards)
+
+	var idx int
```
Comment (raulk) [inline node/impl/full/eth.go:2181]: Also adjusted the logic here so that it's easier to follow.

**Pull request merged**